File: example/go-redis/redis/pub-sub.go
package redis

import (
	"context"
	"fmt"
)

func (r *repository) PUBLISH(ctx context.Context, channel, message string) error {
	return r.client.Publish(
		ctx,
		channel,
		message,
	).Err()
}

func (r *repository) SUBSCRIBE(ctx context.Context, channel string) error {
	pubsub := r.client.Subscribe(ctx, channel)
	ch := pubsub.Channel()
	fmt.Printf("Subscribed to %q\n", channel)
	for {
		select {
		case <-ctx.Done():
			return ctx.Err()

		case msg := <-ch:
			fmt.Printf(
				"Received message from %q: %s\n",
				msg.Channel,
				msg.Payload,
			)
		}
	}
}

File: example/rust-redis/src/redis/pub_sub.rs
use anyhow::Result;
use futures_util::StreamExt;
use redis::AsyncCommands;

use super::redis::Repository;

impl Repository {
    pub async fn publish(
        &self,
        channel: &str,
        message: &str,
    ) -> Result<()> {
        let mut conn = self.client.get_multiplexed_async_connection().await?;

        conn.publish::<_, _, ()>(
            channel,
            message,
        )
        .await?;

        Ok(())
    }

    pub async fn subscribe(
        &self,
        channel: &str,
    ) -> Result<()> {
        let mut pubsub = self.client.get_async_pubsub().await?;

        pubsub.subscribe(channel).await?;

        println!("Subscribed to \"{}\"", channel);

        let mut messages = pubsub.on_message();

        while let Some(message) = messages.next().await {
            let payload: String = message.get_payload()?;

            println!(
                "Received message from \"{}\": {}",
                message.get_channel_name(),
                payload,
            );
        }

        Ok(())
    }
}